### PR TITLE
Structured integer fuse

### PR DESCRIPTION
**Pull request by iceslice0**
Structurally merge linear path to implement integer convolution and lossless BN fuse and complete integer inference 

Work is almost complete except
1. Zero Point Guard Bit parameters (currently hardcoded)
2. Decoupling of quantization and de-quantization in NoisyActLin.forward

Even at this point BN fuse loss < 0.3% accuracy on rensnet20+cifar100 w1a1

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -106,25 +106,25 @@ flake8 your_module/
...
+|   +-- models    # for models definition
+|   |   +-- compose
+|   |       +-- vision
+|   +-- quantization # for quantization methods definition
+|   |   +-- abc
+|   |   +-- dummy
+|   |   +-- gdnsq
+|   +-- training    # for specific training regimes
++-- tests           # for tests
 ```
```
Comment (janchk) [inline CONTRIBUTING.md:152]: Leave my unicode alone.

File: src/quantization/gdnsq/gdnsq_loss.py [hunk 2]
```diff
@@ -85,84 +103,6 @@ def forward(self, output, target):
...
-        self.rloss = rloss
-        self.s_weight_loss = -lws.mean()
-        self.q_weight_loss = lwq.mean()
-        self.s_act_loss = -las.mean()
-        self.q_act_loss = laq.mean()
-        self.weight_reg_loss = (lwq-lws).max()
-
-        return ploss
\ No newline at end of file
+        return super().forward(output, target=None)
```
Comment (janchk) [inline src/quantization/gdnsq/gdnsq_loss.py:108]: Nice one. It was a quite dirty hack before..

File: scripts/gdnsq_q_validate.py [hunk 3]
```diff
@@ -17,6 +25,17 @@
...
+    # Prevent construction of WandbLogger inside src/training/trainer.py
+    try:
+        tloggers = config.training.loggers
+        if isinstance(tloggers, dict):
+            tloggers.pop("WandbLogger", None)
+            tloggers.pop("wandb", None)
+            tloggers.pop("wandb_logger", None)
+    except Exception:
+        pass
+
```
Comment (janchk) [inline scripts/gdnsq_q_validate.py]: That's actually very verbose. I'm sure it's cursor's writing. It's can't comprehend all repo's structure and provide local hacks.
Better way is just disable logging for Validator passing "loggers=None". So pipeline would be correct at the stage of construction and not afterwards.

File: scripts/gdnsq_q_predict.py [hunk 4]
```diff
@@ -17,6 +25,17 @@
...
+    # Prevent construction of WandbLogger inside src/training/trainer.py
+    try:
+        tloggers = config.training.loggers
+        if isinstance(tloggers, dict):
+            tloggers.pop("WandbLogger", None)
+            tloggers.pop("wandb", None)
+            tloggers.pop("wandb_logger", None)
+    except Exception:
+        pass
+
```
Comment (janchk) [inline scripts/gdnsq_q_predict.py]: Same excessive "disabling code"

File: scripts/gdnsq_q_fuse_batchnorm.py [hunk 5]
```diff
@@ -0,0 +1,301 @@
...
+
+def _get_model_root(model: torch.nn.Module) -> torch.nn.Module:
+    return model.model if hasattr(model, "model") else model
+
+
+def materialize_exact_integer_convs_no_batchnorm_fuse(model: torch.nn.Module) -> int:
+    """
+    Replace each ``NoisyConv2d`` with ``ExactIntegerConv2d`` **without** BN folding.
+
+    Used only by ``gdnsq_q_binarize_edges.py``: it reloads the ``state_dict`` from
```
Comment (janchk) [inline scripts/gdnsq_q_fuse_batchnorm.py]: What is ``gdnsq_q_binarize_edges.py`` file?

File: scripts/gdnsq_q_fuse_batchnorm.py [hunk 6]
```diff
@@ -0,0 +1,301 @@
...
+    logger.info(f"Model after layer replacement:\n{qmodel}")
+
+    ckpt = torch.load(args.checkpoint, map_location="cpu", weights_only=False)
+    state = ckpt.get("state_dict", ckpt)
+    qmodel.load_state_dict(state, strict=False)
+    qmodel.eval()
+
+    use_cuda = torch.cuda.is_available()
+    if use_cuda:
+        qmodel = qmodel.to("cuda")
```
Comment (janchk) [inline scripts/gdnsq_q_fuse_batchnorm.py]: The heck is that? Torch lightning do the placements of model and data automatically. *I guess cursor do not know that.*

File: scripts/gdnsq_q_fuse_batchnorm.py [hunk 7]
```diff
@@ -0,0 +1,301 @@
...
+def _disable_tracing(config):
+    # Prevent construction of WandbLogger inside src/training/trainer.py
+    try:
+        tloggers = config.training.loggers
+        if isinstance(tloggers, dict):
+            tloggers.pop("WandbLogger", None)
+            tloggers.pop("wandb", None)
+            tloggers.pop("wandb_logger", None)
+    except Exception:
+        pass
```
Comment (janchk) [inline scripts/gdnsq_q_fuse_batchnorm.py]: Why not disable logging in `_config_for_fuse_validation` since you are override the config anyway..

File: scripts/gdnsq_q_fuse_batchnorm.py [hunk 8]
```diff
@@ -0,0 +1,301 @@
+"""
+Fuse BatchNorm into the previous NoisyConv2d, normalize activation scales,
+run validation, print weight/bias value-set stats, and save fused checkpoint.
+
+The file passed to ``--output`` is the required ``--checkpoint`` for
+``scripts/gdnsq_q_binarize_edges.py`` (same YAML ``--config``).
+"""
```
Comment (janchk) [inline scripts/gdnsq_q_fuse_batchnorm.py]: Thanks to the cpt obvious..

**Review by janchk — COMMENTED** [review]
I didn't like `gdnsq_q_fuse_batchnorm.py` not even a little. As well as the wandb hacks. I'll rewrite them in the following pr to this branch. Otherwise it's cool if it's working. I would advise you to post updated metrics numbers into `README.md`  file.

**Review by janchk — CHANGES_REQUESTED** [review]
I didn't like gdnsq_q_fuse_batchnorm.py not even a little. As well as the wandb hacks. I'll rewrite them in the following pr to this branch. Otherwise it's cool if it's working. I would advise you to post updated metrics numbers into README.md file.

**Review by janchk — APPROVED** [review]
Good
